File: Java/Java Project/Iteration_2/src/FileManager1.java
package Iteration_2.src;


import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;
import org.apache.log4j.Logger;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.Reader;
import java.lang.reflect.Type;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.HashMap;

public class FileManager1 {



    public ArrayList<Student> readStudent(String path) {
        Logger logger = Logger.getLogger(FileManager1.class.getName());
        ArrayList<Student> students = new ArrayList<>();
        Gson gson = new GsonBuilder().setPrettyPrinting().setLenient().create();
        File directoryPath = new File(path);
        File[] fileList = directoryPath.listFiles();

        for (File file :fileList) {
            try {

                Reader reader = Files.newBufferedReader(Path.of(file.getAbsolutePath()));
                Type studentsType = new TypeToken<Student>() {
                }.getType();
                students.add(gson.fromJson(reader, studentsType));
            } catch (IOException e) {
                e.printStackTrace();
            }

        }
        logger.info(students.size()+ " Student objects created.");
        return students;
    }

    public ArrayList<CompulsoryCourse> readCourse(String path) {
        try {
            Gson gson = new Gson();
            Reader reader = Files.newBufferedReader(Path.of(path));
            Type courseType = new TypeToken<ArrayList<CompulsoryCourse>>() {
            }.getType();

            return gson.fromJson(reader, courseType);

        } catch (IOException e) {
            e.printStackTrace();
        }
        return null;
    }
    public ArrayList<Student> readStudent1(String path) {
        try {
            Gson gson = new Gson();
            Reader reader = Files.newBufferedReader(Path.of(path));
            Type studentType = new TypeToken<ArrayList<Student>>() {
            }.getType();
            return gson.fromJson(reader, studentType);

        } catch (IOException e) {
            e.printStackTrace();
        }
        return null;
    }
    public ArrayList<Lecturer> readLecturers(String path) {
        Logger logger = Logger.getLogger(FileManager1.class.getName());
        try {
            Gson gson = new Gson();
            Reader reader = Files.newBufferedReader(Path.of(path));
            Type lecturerType = new TypeToken<ArrayList<Lecturer>>() {
            }.getType();

            ArrayList<Lecturer> lecturers = gson.fromJson(reader, lecturerType);
            logger.info(lecturers.size()+ " lecturer is read and lecturer objects created.");
            return lecturers;

        } catch (IOException e) {
            e.printStackTrace();
        }
        return null;
    }


    public HashMap<String, String> readParams(String path) {
        Logger logger = Logger.getLogger(FileManager1.class.getName());
        try {
            Gson gson = new Gson();
            Reader reader = Files.newBufferedReader(Path.of(path));
            Type classType = new TypeToken<HashMap<String,String>>(){}.getType();
            logger.info("Parameters are read and initialized. ");
            return gson.fromJson(reader, classType);

        } catch (IOException e) {
            e.printStackTrace();
        }
        return null;

    }

    public void writeToFile(Student student, String filePath) throws IOException {
         Logger logger = Logger.getLogger(FileManager1.class.getName());
        try {
            File file = new File(filePath);
            Gson gson = new GsonBuilder().setPrettyPrinting().serializeSpecialFloatingPointValues().create();
            FileWriter fileWriter = new FileWriter(file, false);
            fileWriter.write(gson.toJson(student).toString());
            fileWriter.close();


        } catch (IOException e) {
            e.printStackTrace();
        }
        logger.info(student.getStudentID()+".json is created. ");

    }

    public void writeToFileWithArray(ArrayList<Student> student, String filePath) throws IOException {
        try {
            File file = new File(filePath);
            Gson gson = new GsonBuilder().setPrettyPrinting().serializeSpecialFloatingPointValues().create();
            FileWriter fileWriter = new FileWriter(file, false);
            gson.toJson(student,fileWriter);
            fileWriter.close();

        } catch (IOException e) {
            e.printStackTrace();
        }
    }

}

File: Python/src/FileManager1.py
import json
import logging
from typing import List, Type, TypeVar, Any

T = TypeVar('T')

class FileManager1:
    def read_student(self, path: str) -> List[T]:
        logger = logging.getLogger(FileManager1.__name__)
        students = []
        gson = json.JSONEncoder(indent=2, sort_keys=True)
        directory_path = path
        file_list = directory_path.listdir()

        for file in file_list:
            try:
                with open(file.abs_path, 'r') as reader:
                    students_type = Type[T]
                    students.append(gson.from_json(reader, students_type))
            except Exception as e:
                print(e)

        logger.info(f'{len(students)} Student objects created.')
        return students

    def read_course(self, path: str) -> List[T]:
        try:
            gson = json.JSONEncoder(indent=2, sort_keys=True)
            with open(path, 'r') as reader:
                course_type = Type[List[T]]
                return gson.from_json(reader, course_type)
        except Exception as e:
            print(e)
        return None

    def read_student1(self, path: str) -> List[T]:
        try:
            gson = json.JSONEncoder(indent=2, sort_keys=True)
            with open(path, 'r') as reader:
                student_type = Type[List[T]]
                return gson.from_json(reader, student_type)
        except Exception as e:
            print(e)
        return None

    def read_lecturers(self, path: str) -> List[T]:
        logger = logging.getLogger(FileManager1.__name__)
        try:
            gson = json.JSONEncoder(indent=2, sort_keys=False)
            with open(path, 'r') as reader:
                lecturer_type = Type[List[T]]
                lecturers = gson.from_json(reader, lecturer_type)
                logger.info(f'{len(lecturers)} lecturer is read and lecturer objects created.')
                return lecturers
        except Exception as e:
            print(e)
        return None

    def read_params(self, path: str) -> dict[str, str]:
        logger = logging.getLogger(FileManager1.__name__)
        try:
            gson = json.JSONEncoder(indent=2, sort_keys=True)
            with open(path, 'r') as reader:
                class_type = Type[dict[str, str]]
                logger.info('Parameters are read and initialized.')
                return gson.from_json(reader, class_type)
        except Exception as e:
            print(e)
        return None

    def write_to_file(student: T, file_path: str) -> None:
        logger = logging.getLogger(FileManager1.__name__)
        try:
            file = file_path
            gson = json.JSONEncoder(indent=2, sort_keys=True, ensure_ascii=True)
            with open(file, 'w') as file_writer:
                json.dump(student, file_writer, default=lambda o: o.__dict__)
        except Exception as e:
            print(e)
        logger.info(f'{student.student_id}.json is created.')

    def write_to_file_with_array(student: List[T], file_path: str) -> None:
        try:
            file = file_path
            gson = json.JSONEncoder(indent=2, sort_keys=True, ensure_ascii=True)
            with open(file, 'w') as file_writer:
                json.dump(student, file_writer, default=lambda o: o.__dict__)
        except Exception as e:
            print(e)
